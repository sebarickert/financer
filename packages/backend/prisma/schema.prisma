generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

enum Role {
  ADMIN
  TEST_USER
}

enum AccountType {
  CASH
  SAVINGS
  INVESTMENT
  CREDIT
  LOAN
  LONG_TERM_SAVINGS
  PRE_ASSIGNED_CASH
}

enum TransactionType {
  INCOME
  EXPENSE
  TRANSFER
}

enum TransactionTemplateType {
  MANUAL
  AUTO
}

enum UserPreferenceProperty {
  DEFAULT_INCOME_ACCOUNT
  DEFAULT_EXPENSE_ACCOUNT
  DEFAULT_TRANSFER_SOURCE_ACCOUNT
  DEFAULT_TRANSFER_TARGET_ACCOUNT
  TRANSACTION_LIST_CHUNK_SIZE
  UPDATE_INVESTMENT_MARKET_VALUE
  DASHBOARD_SETTINGS
  STATISTICS_SETTINGS
}

model AccountBalanceChange {
  id        String   @id @default(auto()) @map("_id") @db.ObjectId
  v         Int?     @map("__v")
  accountId String   @db.ObjectId
  amount    Float
  createdAt DateTime @default(now()) @db.Date
  date      DateTime @db.Date
  updatedAt DateTime @updatedAt @db.Date
  userId    String   @db.ObjectId

  account Account @relation(fields: [accountId], references: [id])
  user    User    @relation(fields: [userId], references: [id])

  @@map("account-balance-changes")
}

model Account {
  id        String      @id @default(auto()) @map("_id") @db.ObjectId
  v         Int?        @map("__v")
  balance   Float
  createdAt DateTime    @default(now()) @db.Date
  isDeleted Boolean
  name      String
  userId    String      @map("owner") @db.ObjectId
  type      AccountType
  updatedAt DateTime    @updatedAt @db.Date

  user                 User                   @relation(fields: [userId], references: [id])
  AccountBalanceChange AccountBalanceChange[]
  fromTransactions     Transaction[]          @relation("FromAccountTransaction")
  toTransactions       Transaction[]          @relation("ToAccountTransaction")
  fromTemplates        TransactionTemplate[]  @relation("FromAccountTemplate")
  toTemplates          TransactionTemplate[]  @relation("ToAccountTemplate")

  @@index([type], map: "type_1")
  @@index([userId], map: "owner_1")
  @@index([isDeleted], map: "isDeleted_1")
  @@map("accounts")
}

model sessions {
  id           String   @id @map("_id")
  expires      DateTime @db.Date
  lastModified DateTime @db.Date
  session      String

  @@index([expires], map: "expires_1")
}

model systemlog {
  id String @id @default(auto()) @map("_id") @db.ObjectId
}

model TransactionCategory {
  id               String            @id @default(auto()) @map("_id") @db.ObjectId
  v                Int?              @map("__v")
  createdAt        DateTime          @default(now()) @db.Date
  deleted          Boolean
  name             String
  userId           String            @map("owner") @db.ObjectId
  parentCategoryId String?           @map("parent_category_id") @db.ObjectId
  updatedAt        DateTime          @updatedAt @db.Date
  visibility       TransactionType[]

  user                       User                         @relation(fields: [userId], references: [id])
  // category TransactionCategory @relation((fields: [parentCategoryId], references: [id]))
  TransactionCategoryMapping TransactionCategoryMapping[]

  @@index([userId], map: "owner_1")
  @@index([visibility], map: "visibility_1")
  @@index([deleted], map: "deleted_1")
  @@map("transaction-categories")
}

model TransactionCategoryMapping {
  id            String   @id @default(auto()) @map("_id") @db.ObjectId
  v             Int?     @map("__v")
  amount        Float
  categoryId    String   @map("category_id") @db.ObjectId
  createdAt     DateTime @default(now()) @db.Date
  description   String
  userId        String   @map("owner") @db.ObjectId
  transactionId String   @map("transaction_id") @db.ObjectId
  updatedAt     DateTime @updatedAt @db.Date

  user        User                @relation(fields: [userId], references: [id])
  category    TransactionCategory @relation(fields: [categoryId], references: [id])
  transaction Transaction         @relation(fields: [transactionId], references: [id])

  @@index([userId], map: "owner_1")
  @@index([transactionId], map: "transaction_id_1")
  @@map("transaction-category-mappings")
}

model TransactionTemplateLog {
  id            String                  @id @default(auto()) @map("_id") @db.ObjectId
  eventType     TransactionTemplateType
  executed      DateTime                @db.Date
  templateId    String                  @db.ObjectId
  transactionId String                  @db.ObjectId
  userId        String                  @db.ObjectId

  user        User                @relation(fields: [userId], references: [id])
  transaction Transaction         @relation(fields: [transactionId], references: [id])
  template    TransactionTemplate @relation(fields: [templateId], references: [id])

  @@index([userId], map: "userId_1")
  @@index([eventType], map: "eventType_1")
  @@index([transactionId], map: "transactionId_1")
  @@index([templateId], map: "templateId_1")
  @@index([executed], map: "executed_1")
  @@map("transaction-template-logs")
}

model TransactionTemplate {
  id                 String                    @id @default(auto()) @map("_id") @db.ObjectId
  v                  Int?                      @map("__v")
  amount             Float
  categories         String[]                  @default([])
  createdAt          DateTime                  @default(now()) @db.Date
  dayOfMonth         Int?
  dayOfMonthToCreate Int?
  description        String
  fromAccount        String?                   @db.ObjectId
  templateName       String
  templateType       TransactionTemplateType[]
  templateVisibility TransactionType?
  toAccount          String?                   @db.ObjectId
  updatedAt          DateTime                  @updatedAt @db.Date
  userId             String                    @db.ObjectId

  user                   User                     @relation(fields: [userId], references: [id])
  from                   Account?                 @relation("FromAccountTemplate", fields: [fromAccount], references: [id])
  to                     Account?                 @relation("ToAccountTemplate", fields: [toAccount], references: [id])
  TransactionTemplateLog TransactionTemplateLog[]

  @@index([userId], map: "userId_1")
  @@index([templateType], map: "templateType_1")
  @@index([templateVisibility], map: "templateVisibility_1")
  @@index([dayOfMonthToCreate], map: "dayOfMonthToCreate_1")
  @@map("transaction-templates")
}

model Transaction {
  id          String   @id @default(auto()) @map("_id") @db.ObjectId
  v           Int?     @map("__v")
  amount      Float
  createdAt   DateTime @default(now()) @db.Date
  date        DateTime @db.Date
  description String
  fromAccount String?  @db.ObjectId
  toAccount   String?  @db.ObjectId
  updatedAt   DateTime @updatedAt @db.Date
  userId      String   @map("user") @db.ObjectId

  user                       User                         @relation(fields: [userId], references: [id])
  from                       Account?                     @relation("FromAccountTransaction", fields: [fromAccount], references: [id])
  to                         Account?                     @relation("ToAccountTransaction", fields: [toAccount], references: [id])
  TransactionCategoryMapping TransactionCategoryMapping[]
  TransactionTemplateLog     TransactionTemplateLog[]

  @@index([fromAccount], map: "fromAccount_1")
  @@index([toAccount], map: "toAccount_1")
  @@index([date], map: "date_1")
  @@index([userId], map: "user_1")
  @@map("transactions")
}

model UserPreferences {
  id        String                 @id @default(auto()) @map("_id") @db.ObjectId
  v         Int?                   @map("__v")
  createdAt DateTime               @default(now()) @db.Date
  key       UserPreferenceProperty
  updatedAt DateTime               @updatedAt @db.Date
  userId    String                 @db.ObjectId
  value     String

  user User @relation(fields: [userId], references: [id])

  @@index([userId], map: "userId_1")
  @@map("userpreferences")
}

model User {
  id                         String                       @id @default(auto()) @map("_id") @db.ObjectId
  v                          Int?                         @map("__v")
  auth0Id                    String?                      @unique
  createdAt                  DateTime                     @default(now()) @db.Date
  githubId                   String?                      @unique
  name                       String
  nickname                   String
  profileImageUrl            String
  roles                      Role[]                       @default([])
  updatedAt                  DateTime                     @updatedAt @db.Date
  AccountBalanceChange       AccountBalanceChange[]
  Account                    Account[]
  TransactionCategory        TransactionCategory[]
  TransactionCategoryMapping TransactionCategoryMapping[]
  TransactionTemplateLog     TransactionTemplateLog[]
  TransactionTemplate        TransactionTemplate[]
  Transaction                Transaction[]
  UserPreferences            UserPreferences[]

  @@map("users")
}
