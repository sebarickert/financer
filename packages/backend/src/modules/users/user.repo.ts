import { Injectable } from '@nestjs/common';
import { Prisma, User } from '@prisma/client';

import { PrismaService } from '@/database/prisma.service';

@Injectable()
export class UserRepo {
  constructor(private readonly prisma: PrismaService) {}

  async findOne(
    userWhereUniqueInput: Prisma.UserWhereUniqueInput,
  ): Promise<User | null> {
    return this.prisma.user.findUnique({
      where: userWhereUniqueInput,
    });
  }

  async findMany(params: {
    skip?: number;
    take?: number;
    cursor?: Prisma.UserWhereUniqueInput;
    where?: Prisma.UserWhereInput;
    orderBy?: Prisma.UserOrderByWithRelationInput;
  }): Promise<User[]> {
    const { skip, take, cursor, where, orderBy } = params;
    return this.prisma.user.findMany({
      skip,
      take,
      cursor,
      where,
      orderBy,
    });
  }

  async create(data: Prisma.UserCreateInput): Promise<User> {
    const { id, email, name, password, createdAt, updatedAt } = data;
    return this.prisma.user.create({
      data: {
        id,
        email,
        name,
        password,
        createdAt,
        updatedAt,
      },
    });
  }

  async update(params: {
    where: Prisma.UserWhereUniqueInput;
    data: Prisma.UserUpdateInput;
  }): Promise<User> {
    const { where, data } = params;
    const { id, email, name, password, createdAt, updatedAt } = data as Prisma.UserUncheckedUpdateInput;
    return this.prisma.user.update({
      data: {
        id,
        email,
        name,
        password,
        createdAt,
        updatedAt,
      },
      where,
    });
  }

  async delete(where: Prisma.UserWhereUniqueInput): Promise<User> {
    return this.prisma.user.delete({
      where,
    });
  }
}
