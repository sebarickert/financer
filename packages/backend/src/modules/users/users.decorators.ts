import { createParamDecorator, ExecutionContext } from '@nestjs/common';
import { User } from '@prisma/client';

import { parseObjectId } from '../../types/objectId';

export const UserId = createParamDecorator(
  (data: unknown, ctx: ExecutionContext) => {
    const request = ctx.switchToHttp().getRequest();
    const user = request.user as User;

    return user.id;
  },
);

export const UserIdOld = createParamDecorator(
  (data: unknown, ctx: ExecutionContext) => {
    const request = ctx.switchToHttp().getRequest();
    const user = request.user as User;

    return parseObjectId(user.id);
  },
);
