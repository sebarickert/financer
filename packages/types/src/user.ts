export enum Role {
  admin = 'admin',
  testUser = 'test-user',
}

export type IUser = {
  _id: any;
  name: any;
  nickname: string;
  githubId?: string;
  auth0Id?: string;
  profileImageUrl: string;
  roles: Role[];
};
