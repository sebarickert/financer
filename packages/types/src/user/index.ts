export * from './user.dto';
export * from './role';
export * from './create-user.dto';
export * from './update-user.dto';
