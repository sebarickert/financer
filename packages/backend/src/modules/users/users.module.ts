import { forwardRef, Module } from '@nestjs/common';

import { DatabaseModule } from '../../database/prisma.module';
import { UserDataModule } from '../user-data/user-data.module';

import { UsersController } from './users.controller';
import { UsersService } from './users.service';

@Module({
  imports: [DatabaseModule, forwardRef(() => UserDataModule)],
  controllers: [UsersController],
  providers: [UsersService],
  exports: [UsersService],
})
export class UsersModule {}
