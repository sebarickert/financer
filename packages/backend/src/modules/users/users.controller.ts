import { Role } from '@local/types';
import {
  Controller,
  Get,
  Body,
  Patch,
  Param,
  Post,
  Res,
  forwardRef,
  Inject,
} from '@nestjs/common';
import {
  ApiBody,
  ApiOkResponse,
  ApiParam,
  ApiTags,
} from '@silte/nestjs-swagger';
import { Response } from 'express';

import { ObjectId } from '../../types/objectId';
import { ValidateEntityId } from '../../utils/validate-entity-id.pipe';
import { Auth } from '../auth/decorators/auht.decorator';
import { LoggedIn } from '../auth/decorators/loggedIn.decorators';
import { UserDataExportDto } from '../user-data/dto/user-data-export.dto';
import { UserDataImportDto } from '../user-data/dto/user-data-import.dto';
import {
  UserDataService,
  ImportUserDataDto,
} from '../user-data/user-data.service';

import { UpdateUserDto, UpdateUserOwnUserDto } from './dto/update-user.dto';
import { UserDto } from './dto/user.dto';
import { UserId } from './users.decorators';
import { UsersService } from './users.service';

@Controller('api/user')
@LoggedIn()
@ApiTags('Users')
export class UserController {
  constructor(
    private readonly usersService: UsersService,
    @Inject(forwardRef(() => UserDataService))
    private readonly userDataService: UserDataService,
  ) {}

  @Get('my-user')
  @ApiOkResponse({ type: UserDto })
  async findOwnUser(@UserId() userId: string) {
    return this.usersService.findOne(userId);
  }

  @Get('my-user/my-data')
  @ApiOkResponse({ type: UserDataExportDto })
  getAllOwnUserData(@UserId() userId: string, @Res() res: Response) {
    this.getAllOneUserData(userId, res);
  }

  @Post('my-user/my-data')
  @Auth(Role.testUser)
  @ApiBody({ type: UserDataImportDto })
  @ApiOkResponse({ schema: { properties: { payload: { type: 'string' } } } })
  overrideAllOwnUserData(
    @UserId() userId: ObjectId,
    @Body() userData: ImportUserDataDto,
  ) {
    return this.userDataService.overrideUserData(userId, userData);
  }

  @Patch('my-user')
  @ApiBody({ type: UpdateUserOwnUserDto })
  @ApiOkResponse({ type: UserDto })
  updateOwnUser(
    @UserId() userId: string,
    @Body() updateUserDto: UpdateUserOwnUserDto,
  ) {
    return this.usersService.update(userId, updateUserDto);
  }

  @Get()
  @Auth(Role.admin)
  @ApiOkResponse({ type: [UserDto] })
  findAll() {
    return this.usersService.findAll();
  }

  @Get(':id')
  @Auth(Role.admin)
  @ApiOkResponse({ type: UserDto })
  @ApiParam({
    name: 'id',
    type: String,
    description: 'Entity id from users collection.',
  })
  findOne(@Param('id', ValidateEntityId) id: string) {
    return this.usersService.findOne(id);
  }

  @Get(':id/my-data')
  @Auth(Role.admin)
  @ApiOkResponse({ type: UserDataExportDto })
  @ApiParam({
    name: 'id',
    type: String,
    description: 'Entity id from users collection.',
  })
  async getAllOneUserData(
    @Param('id', ValidateEntityId) userId: string,
    @Res() res: Response,
  ) {
    const { filename, data } = await this.userDataService.findAllOneUserData(
      userId,
    );

    res.setHeader('Content-disposition', `attachment; filename= ${filename}`);
    res.setHeader('Content-type', 'application/json');

    res.write(JSON.stringify(data), () => {
      res.end();
    });
  }

  @Patch(':id')
  @Auth('admin')
  @ApiBody({ type: UpdateUserDto })
  @ApiOkResponse({ type: UserDto })
  @ApiParam({
    name: 'id',
    type: String,
    description: 'Entity id from users collection.',
  })
  update(
    @Param('id', ValidateEntityId) id: string,
    @Body() updateUserDto: UpdateUserDto,
  ) {
    return this.usersService.update(id, updateUserDto);
  }
}
