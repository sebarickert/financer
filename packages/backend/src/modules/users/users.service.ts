import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';

import { isNodeEnvInTest } from '../../config/configuration';
import { DUMMY_TEST_USER } from '../../config/mockAuthenticationMiddleware';
import { ObjectId } from '../../types/objectId';

import { UpdateUserDto } from './dto/update-user.dto';
import { User, UserDocument } from './schemas/user.schema';

@Injectable()
export class UsersService {
  constructor(@InjectModel(User.name) private userModel: Model<UserDocument>) {}

  async findAll(): Promise<UserDocument[]> {
    return this.userModel.find().exec();
  }

  async findOne(id: ObjectId): Promise<UserDocument> {
    return !isNodeEnvInTest()
      ? this.userModel.findById(id).exec()
      : (DUMMY_TEST_USER as UserDocument);
  }

  async findOneByGithubId(githubId: string): Promise<UserDocument> {
    return this.userModel.findOne({ githubId }).exec();
  }

  async findOneByAuth0Id(auth0Id: string): Promise<UserDocument> {
    return this.userModel.findOne({ auth0Id }).exec();
  }

  async create(createUserDto: User) {
    return this.userModel.create(createUserDto);
  }

  update(id: ObjectId, updateUserDto: UpdateUserDto) {
    return this.userModel.findByIdAndUpdate(id, updateUserDto, { new: true });
  }

  remove(id: ObjectId) {
    return `This action removes a #${id} user`;
  }
}
