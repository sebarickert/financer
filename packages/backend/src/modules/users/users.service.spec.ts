import { MongooseModule } from '@nestjs/mongoose';
import { Test, TestingModule } from '@nestjs/testing';

import { rootMongooseTestModule } from '../../../test/rootMongooseTest.module';

import { User, UserSchema } from './schemas/user.schema';
import { UsersService } from './users.service';

describe('UsersService', () => {
  let service: UsersService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        rootMongooseTestModule(),
        MongooseModule.forFeature([{ name: User.name, schema: UserSchema }]),
      ],
      providers: [UsersService],
    }).compile();

    service = module.get<UsersService>(UsersService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
});
