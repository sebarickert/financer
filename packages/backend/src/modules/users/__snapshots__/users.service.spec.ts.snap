// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersService should return a user by id 1`] = `
{
  "auth0Id": "",
  "githubId": "",
  "id": "61460d7354ea082ad0256749",
  "name": "Dummy Test User",
  "nickname": "",
  "profileImageUrl": "",
  "roles": [
    "TEST_USER",
  ],
  "theme": "AUTO",
}
`;

exports[`UsersService should return an array of users from findAll 1`] = `
[
  {
    "auth0Id": "",
    "githubId": "",
    "id": "61460d7354ea082ad0256749",
    "name": "Dummy Test User",
    "nickname": "",
    "profileImageUrl": "",
    "roles": [
      "TEST_USER",
    ],
    "theme": "AUTO",
  },
]
`;
